src/codegen/subexpression_elimination/anticipated_expressions.rs

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::subexpression_elimination::{
+    kill_loop_variables, AvailableExpression, AvailableExpressionSet,
+};
+use crate::codegen::Expression;
+use std::collections::HashMap;
+
+/// The AnticipatedExpression struct manages everything related to traversing the CFG backwards, so
+/// we can perform an anticipated expression analysis.
+///
+/// "An expression is anticipated at a point if it is certain to be evaluated along any
+/// path before this expression's value is changed (any variables its evaluation depends
+/// on are reassigned)."
+///
+/// The anticipated expression analysis helps us find where a common subexpression can be evaluated.
```

Comment:
This doesn't describe anticipated expression. How about:

Available expressions shows us where sub expressions are already evaluated. Using anticipated expression we can do one more step: determine when an expression can be evaluated, because all the definitions it depends on are available. 

This is can be used for moving a common subexpression to another block.

Reply:
You could include a reference to the compiler book chapter on anticipated expressions